### PR TITLE
Adding a 2D backward-facing step example to documentation

### PR DESCRIPTION
**Pull request by charles-lp**
This pull request adds a 2D backward-facing step example to the official documentation. The following documents are added : 

- 2d-backward-facing-step.rst and associated images directory
- [modified] incompressible-flow.rst (Added new entry)
- backward_facing_step.geo
- backward_facing_step.msh
- 2D_backward_facing_step_steady.prm
- 2D_backward_facing_step_steadybdf.prm

### CODE REVIEW

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 1]
```diff
@@ -0,0 +1,348 @@
+====================================
+2D Flow Past a Backward-Facing Step  
+====================================
+
+Features
+--------
+
+- Solver : ``gls_navier_stokes_2d`` 
+- Steady and pseudo steady state solution
+- Comparison with benchmark and analytical data
```
Comment (blaisb) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:10]:
```suggestion
- Comparison with benchmark solutions 
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+In this example, a bidimensional flow goes past a backward-facing step. The flow enters from the left inlet and seperates from the bottom wall at the step, and then reattaches to it further downstream at a distance :math:`x_r` from the step.  
+
+.. image:: image/backward_facing_step_description.png
+
+The backward-facing step problem is a classical computational fluid dynamics problem. The fact that it features a non-trivial solution while maintaining simple geometry and boundary conditions makes this problem a good candidate for validation purposes as well as to test the robustness of a given CFD method. First, the basic parameters used to solve the backward-facing step problem will be exposed. A solution to several Reynolds numbers (from :math:`Re = 100` to :math:`Re =1000`) will then be presented and compared to experimental and numerical data. A mesh adaptation and numerical error analysis will be carried on.
+
+Parameter File
+--------------
+
+The following subsections show the different parameters of *Lethe* used in the simulations. While they all remain more or less the same throughout the various cases, some of them change as the Reynolds number is increased.
```
Comment (blaisb) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:32]:
```suggestion
The following subsections show the different parameters used in the simulations. While they all remain more or less the same throughout the various cases, some of them change as the Reynolds number is increased.
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+	  set number mesh adapt = 10
+	  set output name = backward_facing_step_output
+	  set output frequency = 1
+	  set subdivision = 1
+	  set output boundaries = false
+	end
+	
+A mesh refinement analysis can be done with ``set number mesh adapt = 10``. By starting from a very coarse mesh and by dynamically refining the mesh at least 10 times, asymptotic convergence can be clearly observed.
+
+However, for :math:`Re \geq 700`, convergence can be quite difficult to obtain while doing a steady state simulation. In fact, as the Reynolds number increases, the problem becomes progressively stiffer to a point where the ``steady`` solver ultimately fails. With that in mind, the case can be solved as a transient problem until the steady state solution is obtained. This can be achieved with the ``method = steady_bdf`` parameter.
```
Comment (blaisb) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:52]: Well-written, def. you are good at this!

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+	  set max cfl                       = 1e6
+	  set adaptative time step scaling = 1.2
+	  set number mesh adapt = 0
+	  set output name = backward_facing_step_output
+	  set output frequency = 1
+	  set subdivision = 1
+	  set output boundaries = false
+	end
+  
+``stop tolerance``, ``time step``, ``adapt``, ``max cfl`` and ``adaptive time step scaling`` are parameters that control the pseudo-steady simulation. Since this problem can be quite demanding, a relatively high tolerance has been chosen as an input for ``stop tolerance``. This choice is however not without its consequences : the :math:`x_r` parameter might be slightly underestimated.
```
Comment (blaisb) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:70]: Have you checked if changing this parameter changed x_r? 1e-6 is actually already quite low.

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+~~~~
+
+.. code-block:: text
+
+	subsection mesh
+	  set type = gmsh
+	  set file name = ../backward_facing_step.msh
+	end
+	
+The mesh features quad elements as well as unit step and inlet heights (:math:`h_{in}=h=1`). In that direction, the expansion ratio has been set to :math:`\beta=\frac{h_{out}}{h_{in}}=2` throughout the entirety of the simulations. Also, the inlet and outlet lengths should be long enough that they allow the formation of a fully developed flow and their respective end. Also, since a gmsh mesh file is used, the initial mesh should be as coarse, since these cells cannot be coarsened with the mesh adaptation algorithm.
```
Comment (blaisb) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:111]: What do you mean by "their respective end"

**Review by blaisb — COMMENTED** [review]
Good work. Only a few comments there and there. Nice job seriously! I am amazed you managed to build this up from pretty much zero! Nice work!

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+-----------------------
+
+- Base case (:math:`Re < 700`) : ``/examples/incompressible_flow/2d-backward-facing-step/Reynolds100-600``
+- Higher-Reynolds case (:math:`Re \geq 700`) : ``/examples/incompressible_flow/2d-backward-facing-step/Reynolds700-1000``
+
+
+Description of the case
+-----------------------
+
+In this example, a bidimensional flow goes past a backward-facing step. The flow enters from the left inlet and seperates from the bottom wall at the step, and then reattaches to it further downstream at a distance :math:`x_r` from the step.
```
Comment (CaroleAnneDaunais) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:23]:
```suggestion
In this example, a bidimensional flow goes past a backward-facing step. The flow enters from the left inlet and separates from the bottom wall at the step, and then reattaches to it further downstream at a distance :math:`x_r` from the step.  
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst (hunk 5 above)
Comment (CaroleAnneDaunais) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:111]:
```suggestion
The mesh features quad elements as well as unit step and inlet heights (:math:`h_{in}=h=1`). In that direction, the expansion ratio has been set to :math:`\beta=\frac{h_{out}}{h_{in}}=2` throughout the entirety of the simulations. Also, the inlet and outlet lengths should be long enough that they allow the formation of a fully developed flow and their respective end. Also, since a ``gmsh`` mesh file is used, the initial mesh should be as coarse, since these cells cannot be coarsened with the mesh adaptation algorithm.
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
++---------------+----------------+----------------+
+| Order 2       | :math:`10.250` | :math:`0.96` % |       
++---------------+----------------+----------------+
+| Erturk (2008) | :math:`10.349` | N/A            |
++---------------+----------------+----------------+
+
+Velocity distribution
+~~~~~~~~~~~~~~~~~~~~~
+
+To validate the quality of the mesh/geometry as well as, it is interesting to compare the obtained velocity distributions with analytical data. The following figures illustrate the velocity distributions at the outlet (right wall) in comparison to the analytical solution.
```
Comment (CaroleAnneDaunais) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:317]:
```suggestion
To validate the quality of the mesh/geometry as well, it is interesting to compare the obtained velocity distributions with analytical data. The following figures illustrate the velocity distributions at the outlet (right wall) in comparison to the analytical solution.
```

**Review by CaroleAnneDaunais — APPROVED** [review]
Very nice PR Charles! Awesome example. I think that for the remaining CI error you get, you just need to merge the master branch into yours (or rebase, as you prefer).

**Comment by blaisb** [comment]
> Very nice PR Charles! Awesome example. I think that for the remaining CI error you get, you just need to merge the master branch into yours (or rebase, as you prefer).

if you need help with that @charles-lp  juste write to me we will do it together.

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+- Solver : ``gls_navier_stokes_2d`` 
+- Steady and pseudo steady state solution
+- Comparison with benchmark and analytical data
+- Mesh refinement and error analysis
+
+Location of the example
+-----------------------
+
+- Base case (:math:`Re < 700`) : ``/examples/incompressible_flow/2d-backward-facing-step/Reynolds100-600``
+- Higher-Reynolds case (:math:`Re \geq 700`) : ``/examples/incompressible_flow/2d-backward-facing-step/Reynolds700-1000``
```
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:17]:
```suggestion
- Base case (:math:`Re < 700`): ``/examples/incompressible_flow/2d-backward-facing-step/Reynolds100-600``
- Higher-Reynolds case (:math:`Re \geq 700`): ``/examples/incompressible_flow/2d-backward-facing-step/Reynolds700-1000``
```

**Review by lpsaavedra — CHANGES_REQUESTED** [review]
Good job! I really like everything you did with this example. I just have a few suggestions and comments :) and welcome to the Lethe family!

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+However, for :math:`Re \geq 700`, convergence can be quite difficult to obtain while doing a steady state simulation. In fact, as the Reynolds number increases, the problem becomes progressively stiffer to a point where the ``steady`` solver ultimately fails. With that in mind, the case can be solved as a transient problem until the steady state solution is obtained. This can be achieved with the ``method = steady_bdf`` parameter.
+
+.. code-block:: text
+
+	subsection simulation control
+	  set method = steady_bdf
+	  set stop tolerance = 1e-6
+	  set time step                     = 0.005
+	  set adapt                         = true
+	  set max cfl                       = 1e6
```
[the remark this answers is not in the document]
Comment (blaisb) [reply doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:61]: Yeah but since it's the steady_bdf solver, that should not be an issue. It is stable for CFL---->\infty

**Comment by charles-lp** [comment]
Thank you for all your comments. The requested changes have been applied and committed.

**Review by lpsaavedra — APPROVED** [review]

**Review by blaisb — APPROVED** [review]
